### ISSUE TITLE
Problems when running multiple batch-calls in parallel

### ISSUE THREAD

**Issue by da-kami** [bitcoindevkit/rust-electrum-client#67]
This is is a bit tricky to describe.

Recently we see quite a lot of these warnings in our logs:

```
2022-05-27 03:32:23  WARN electrum_client::raw_client: got error for req_id 81: JSON(Error("EOF while parsing a value", line: 1, column: 0))
2022-05-27 03:32:23  WARN electrum_client::raw_client: removing all waiting req of this batch
2022-05-27 03:32:23  WARN electrum_client::client: call 'batch_script_get_history' failed with EOF while parsing a value at line 1 column 0, retry: 1/1
```

At first we did not pay much attention to this, but it actually is quite a problem, because the batch processing fails at this point and keeps failing. We just reconnect and keep processing until we run into an error again.

I traced the problem back to this call:

https://github.com/bitcoindevkit/rust-electrum-client/blob/master/src/raw_client.rs#L491

in the `raw_client` code, context:

https://github.com/bitcoindevkit/rust-electrum-client/blob/master/src/raw_client.rs#L479-L495

I found a pattern when this problem happens. It is related to triggering multiple batch-processing calls in parallel.
Whenever a new "group" of a batch is handled as part of one request the next request will receive an empty response from the server. Since the response is literally just `""` (an empty string) the conversion into `serde::Value` will fail.

Here is some analysis from the `electrum_client` `trace` logs:

```
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 77 received a response for its request
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(78)
2022-05-27 03:31:23 TRACE electrum_client::raw_client: <== {"id":78,"jsonrpc":"2.0","result":[]}

2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 78 received a response for its request
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(79)
2022-05-27 03:31:23 TRACE electrum_client::raw_client: <== {"id":79,"jsonrpc":"2.0","result":[]}

2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 79 received a response for its request
2022-05-27 03:31:23 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":80,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(80)
2022-05-27 03:31:23 TRACE electrum_client::raw_client: <== {"id":80,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}

2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 80 received a response for its request
2022-05-27 03:31:36 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":20,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:31:36 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(20)
2022-05-27 03:31:36 TRACE electrum_client::raw_client: <== {"id":20,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}

2022-05-27 03:31:36 TRACE electrum_client::raw_client: Reader thread 20 received a response for its request
2022-05-27 03:31:56 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":21,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:31:56 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(21)
2022-05-27 03:31:56 TRACE electrum_client::raw_client: <== {"id":21,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}

2022-05-27 03:31:56 TRACE electrum_client::raw_client: Reader thread 21 received a response for its request
2022-05-27 03:32:16 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":22,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:32:16 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(22)
2022-05-27 03:32:16 TRACE electrum_client::raw_client: <== {"id":22,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}

2022-05-27 03:32:16 TRACE electrum_client::raw_client: Reader thread 22 received a response for its request
2022-05-27 03:32:23 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":81,"method":"blockchain.scripthash.get_history","params":["78aff35ce7f2a380a6d14a276e20cbc39ead1a974e77789530aa91ce5adf6535"]}
{"jsonrpc":"2.0","id":82,"method":"blockchain.scripthash.get_history","params":["28cdd8261656cc93d90bf0554f13e009f2c688c8c76e5bffc3fca0475060a44c"]}
{"jsonrpc":"2.0","id":83,"method":"blockchain.scripthash.get_history","params":["3bf0d7bec4f8c9d03fe00649136dd39f2d183dbb7259334c189e1423c7b4e9e8"]}
{"jsonrpc":"2.0","id":84,"method":"blockchain.scripthash.get_history","params":["0a689bdf3cc22bacf922b8b770ddf79455189f4dfcf9fe4f162403a360525fb1"]}
{"jsonrpc":"2.0","id":85,"method":"blockchain.scripthash.get_history","params":["d07a040bd65be7a9c85adf3396987eaa10cd59229cbdc188f7657bbfcc0ca366"]}
{"jsonrpc":"2.0","id":86,"method":"blockchain.scripthash.get_history","params":["2db6d12a4664d3da2ff4e58f47f897f02792c783775c3193591f59f4a31ade73"]}
{"jsonrpc":"2.0","id":87,"method":"blockchain.scripthash.get_history","params":["d8c7cecbed27c20a3a0ae27e38a6c382008ed98be55d9702336363d799dcdf65"]}
{"jsonrpc":"2.0","id":88,"method":"blockchain.scripthash.get_history","params":["b82cc9ea44e14ead585e468d060017ff7c3e67210482fec38bdc38568836ba03"]}
{"jsonrpc":"2.0","id":89,"method":"blockchain.scripthash.get_history","params":["4056737d889e085d6e639cde9c98e8a3f318ec7f2d57534e6118449ad8c0d6f0"]}
{"jsonrpc":"2.0","id":90,"method":"blockchain.scripthash.get_history","params":["68178ab6f73c0b800c455206873e53ea51f89424f9e3237ac3deb56a94f1b384"]}
{"jsonrpc":"2.0","id":91,"method":"blockchain.scripthash.get_history","params":["e45ebe3854006b28c25c2113fde8863f7641a9e20055b8447957fa61103cb9ee"]}
{"jsonrpc":"2.0","id":92,"method":"blockchain.scripthash.get_history","params":["4210b957cfee04b255fbe01f153774a0681f8cbaa38f6453eb645157824a270b"]}
{"jsonrpc":"2.0","id":93,"method":"blockchain.scripthash.get_history","params":["31641b026e4a74dee41e74cef8a074bbecf5f42da30a937eb981b47610be89a9"]}
{"jsonrpc":"2.0","id":94,"method":"blockchain.scripthash.get_history","params":["6bc6045776e4f3c71c40bfe1aed5a2574a40fb47dc0d52d5019117f37c92241a"]}
{"jsonrpc":"2.0","id":95,"method":"blockchain.scripthash.get_history","params":["691d1ba1b9a4e19c7101d1a183e96f2a16324a015d4c8266190899b38e8b80bc"]}
{"jsonrpc":"2.0","id":96,"method":"blockchain.scripthash.get_history","params":["d5f14cd379b7d8ae8b63acdbfde8f27a44588899738abf4da86779f4fd296c0f"]}
{"jsonrpc":"2.0","id":97,"method":"blockchain.scripthash.get_history","params":["0edb4e0eea577f5d24265694dffdb2efba5c997fccbb98a7949f7d3f9fdf7211"]}
{"jsonrpc":"2.0","id":98,"method":"blockchain.scripthash.get_history","params":["345856ce9baa2aa4d892655f762978b7590260a8537df177d4dd860c103fbd32"]}
{"jsonrpc":"2.0","id":99,"method":"blockchain.scripthash.get_history","params":["2b38c9fc2f768f625dcdd8b1ff77d5494fcd0087c2829a263e22d5b07238d67f"]}
{"jsonrpc":"2.0","id":100,"method":"blockchain.scripthash.get_history","params":["e6f1f27da20f0d5a52b3edb6eaddd77072223c6b82076dcc89ec9d09f2699a6e"]}

2022-05-27 03:32:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(81)
2022-05-27 03:32:23 TRACE electrum_client::raw_client: <==
2022-05-27 03:32:23  WARN electrum_client::raw_client: got error for req_id 81: JSON(Error("EOF while parsing a value", line: 1, column: 0))
2022-05-27 03:32:23  WARN electrum_client::raw_client: removing all waiting req of this batch
2022-05-27 03:32:23  WARN electrum_client::client: call 'batch_script_get_history' failed with EOF while parsing a value at line 1 column 0, retry: 1/1
```

So, it processes `77`, `79`, `80` 

```
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 77 received a response for its request
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(78)
2022-05-27 03:31:23 TRACE electrum_client::raw_client: <== {"id":78,"jsonrpc":"2.0","result":[]}

2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 78 received a response for its request
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(79)
2022-05-27 03:31:23 TRACE electrum_client::raw_client: <== {"id":79,"jsonrpc":"2.0","result":[]}

2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 79 received a response for its request
2022-05-27 03:31:23 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":80,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:31:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(80)
2022-05-27 03:31:23 TRACE electrum_client::raw_client: <== {"id":80,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}

2022-05-27 03:31:23 TRACE electrum_client::raw_client: Reader thread 80 received a response for its request
2022-05-27 03:31:36 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":20,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:31:36 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(20)
2022-05-27 03:31:36 TRACE electrum_client::raw_client: <== {"id":20,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}
```

then goes on with `20` (of a different batch group as far as I understand the code):

```
2022-05-27 03:31:36 TRACE electrum_client::raw_client: Reader thread 20 received a response for its request
2022-05-27 03:31:56 TRACE electrum_client::raw_client: ==> {"jsonrpc":"2.0","id":21,"method":"blockchain.headers.subscribe","params":[]}
2022-05-27 03:31:56 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(21)
2022-05-27 03:31:56 TRACE electrum_client::raw_client: <== {"id":21,"jsonrpc":"2.0","result":{"height":2250330,"hex":"00008020b0001b91e265af92c54beeb267e71e03c199820a2480d72af9010000000000001984ae45c7d854b9671c1be452d46be2fefd42aff6180b564d62ed34422919308c4590628886021a41cd062b"}}
```

then, when trying to go on with the next of the previous group, `81` we run into the error where the server returns `""` response:

```
2022-05-27 03:32:23 TRACE electrum_client::raw_client: Starting reader thread with `until_message` = Some(81)
2022-05-27 03:32:23 TRACE electrum_client::raw_client: <==
2022-05-27 03:32:23  WARN electrum_client::raw_client: got error for req_id 81: JSON(Error("EOF while parsing a value", line: 1, column: 0))
2022-05-27 03:32:23  WARN electrum_client::raw_client: removing all waiting req of this batch
2022-05-27 03:32:23  WARN electrum_client::client: call 'batch_script_get_history' failed with EOF while parsing a value at line 1 column 0, retry: 1/1
```

This pattern (start processing different batch group, then fail when going on with previous one) is there every time. 
I got more logs if you need - let me know :)

I was unable to get to the bottom of this yet - would be great to get some help from someone that knows the code better.

Note: We are only encountering these problems when running against the public blockstream `electrs-esplora` node (both mainnet and testnet, testnet URL: ssl://blockstream.info:993). When running against my own `electrs` instance I don't run into this. Not sure if this is a coincidence. But, even if this is a server side setting causing this, it is somewhat worrying that batch processing causes this.

**Comment by afilini** [comment]
If you have some time could you try running the same code against other electrum implementations? There's a nice list of servers here, try with electrumx of fulcrum: https://1209k.com/bitcoin-eye/ele.php

I'm also worried this may be a server side issue, worst case we can add code to ignore empty lines

**Comment by da-kami** [comment]
We tried against some other public nodes but were only able to connect to a few. Most of the SSL connections give us errors upon connection. TCP worked with some.

We tried with one node that worked but ran into rate limit issues quite fast, i.e. `Electrum server error: {"code":-101,"message":"excessive resource usage"}`, which makes it hard to reproduce this.

The blockstream servers have the pluspoint that they also run a testnet node, it's hard to find a stable public one out there ;)
Of course we can run our own infrastructure for that, but it's nice to avoid that.

I think at this point it would be nice if the client could ignore empty lines, if you don't mind.
